<a href="https://colab.research.google.com/github/alexcoy06/Data-Science/blob/main/Project%201/Golden-Age.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.

Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [1]:
# importing pandas
import pandas as pd

Next, variables will be created to make the code usable for anyone who wants to try or take a closer look at the following code.

In [2]:
# For those that have either forked or copied my repository.
df_loc = './datasets/movies_and_shows.csv'

# For those that will use the Google Colab link
df_url = 'https://raw.githubusercontent.com/alexcoy06/Data-Science/main/Project%201/datasets/movies_and_shows.csv'

Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [3]:
# Reading the files and storing them to df
# Change this to the appropriate variable per your situation
df = pd.read_csv(df_loc)

Print the first 10 table rows:

In [4]:
# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()
df.head(10)

Unnamed: 0             name                      Character   r0le  \
0           0   Robert De Niro                  Travis Bickle  ACTOR   
1           1     Jodie Foster                  Iris Steensma  ACTOR   
2           2    Albert Brooks                            Tom  ACTOR   
3           3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR   
4           4  Cybill Shepherd                          Betsy  ACTOR   
5           5      Peter Boyle                         Wizard  ACTOR   
6           6   Leonard Harris      Senator Charles Palantine  ACTOR   
7           7   Diahnne Abbott                Concession Girl  ACTOR   
8           8      Gino Ardito             Policeman at Rally  ACTOR   
9           9  Martin Scorsese  Passenger Watching Silhouette  ACTOR   

         TITLE   Type  release Year              genres  imdb sc0re  \
0  Taxi Driver  MOVIE          1976  ['drama', 'crime']         8.2   
1  Taxi Driver  MOVIE          1976  ['drama', 'crime']         8.2   
2  Taxi Driver  MOVIE          1976  ['drama', 'crime']         8.2   
3  Taxi Driver  MOVIE          1976  ['drama', 'crime']         8.2   
4  Taxi Driver  MOVIE          1976  ['drama', 'crime']         8.2   
5  Taxi Driver  MOVIE          1976  ['drama', 'crime']         8.2   
6  Taxi Driver  MOVIE          1976  ['drama', 'crime']         8.2   
7  Taxi Driver  MOVIE          1976  ['drama', 'crime']         8.2   
8  Taxi Driver  MOVIE          1976  ['drama', 'crime']         8.2   
9  Taxi Driver  MOVIE          1976  ['drama', 'crime']         8.2   

   imdb v0tes  
0    808582.0  
1    808582.0  
2    808582.0  
3    808582.0  
4    808582.0  
5    808582.0  
6    808582.0  
7    808582.0  
8    808582.0  
9    808582.0

Obtain the general information about the table with one command:

In [5]:
# obtaining general information about the data in df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    85579 non-null  int64  
 1      name       85579 non-null  object 
 2   Character     85579 non-null  object 
 3   r0le          85579 non-null  object 
 4   TITLE         85578 non-null  object 
 5     Type        85579 non-null  object 
 6   release Year  85579 non-null  int64  
 7   genres        85579 non-null  object 
 8   imdb sc0re    80970 non-null  float64
 9   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(2), object(6)
memory usage: 6.5+ MB


The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'.


### Conclusions <a id='data_review_conclusions'></a>

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [6]:
# the list of column names in the df table
df.columns

Index(['Unnamed: 0', '   name', 'Character', 'r0le', 'TITLE', '  Type',
       'release Year', 'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [7]:
# renaming columns

#df.columns = ['name', 'character', 'role', 'title', 'type', 'release_year', 'genres', 'imdb_score', 'imdb_votes']

# OR/NOTE: https://pandas.pydata.org/docs/reference/series.html
# and https://pandas.pydata.org/docs/reference/frame.html

df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_').str.replace('0', 'o')
df.rename(columns={'   name': 'name'})

unnamed:_o                name                character      role  \
0               0      Robert De Niro            Travis Bickle     ACTOR   
1               1        Jodie Foster            Iris Steensma     ACTOR   
2               2       Albert Brooks                      Tom     ACTOR   
3               3       Harvey Keitel  Matthew 'Sport' Higgins     ACTOR   
4               4     Cybill Shepherd                    Betsy     ACTOR   
...           ...                 ...                      ...       ...   
85574       85574    Adelaida Buscato               Mar??a Paz     ACTOR   
85575       85575  Luz Stella Luengas             Karen Bayona     ACTOR   
85576       85576        In??s Prieto                    Fanny     ACTOR   
85577       85577        Isabel Gaona                   Cacica     ACTOR   
85578       85578      Julian Gaviria                  unknown  DIRECTOR   

             title       type  release_year              genres  imdb_score  \
0      Taxi Driver      MOVIE          1976  ['drama', 'crime']         8.2   
1      Taxi Driver      MOVIE          1976  ['drama', 'crime']         8.2   
2      Taxi Driver      MOVIE          1976  ['drama', 'crime']         8.2   
3      Taxi Driver      MOVIE          1976  ['drama', 'crime']         8.2   
4      Taxi Driver      MOVIE          1976  ['drama', 'crime']         8.2   
...            ...        ...           ...                 ...         ...   
85574      Lokillo  the movie          2021          ['comedy']         3.8   
85575      Lokillo  the movie          2021          ['comedy']         3.8   
85576      Lokillo  the movie          2021          ['comedy']         3.8   
85577      Lokillo      MOVIE          2021          ['comedy']         3.8   
85578      Lokillo  the movie          2021          ['comedy']         3.8   

       imdb_votes  
0        808582.0  
1        808582.0  
2        808582.0  
3        808582.0  
4        808582.0  
...           ...  
85574        68.0  
85575        68.0  
85576        68.0  
85577        68.0  
85578        68.0  

[85579 rows x 10 columns]

Check the result. Print the names of the columns once more:

In [8]:
# checking result: the list of column names
df.columns

Index(['unnamed:_o', 'name', 'character', 'role', 'title', 'type',
       'release_year', 'genres', 'imdb_score', 'imdb_votes'],
      dtype='object')

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [9]:
# calculating missing values
df.isna().sum()

unnamed:_o         0
name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [10]:
# dropping rows where columns with title, scores and votes have missing values
df.dropna()

unnamed:_o                name                character      role  \
0               0      Robert De Niro            Travis Bickle     ACTOR   
1               1        Jodie Foster            Iris Steensma     ACTOR   
2               2       Albert Brooks                      Tom     ACTOR   
3               3       Harvey Keitel  Matthew 'Sport' Higgins     ACTOR   
4               4     Cybill Shepherd                    Betsy     ACTOR   
...           ...                 ...                      ...       ...   
85574       85574    Adelaida Buscato               Mar??a Paz     ACTOR   
85575       85575  Luz Stella Luengas             Karen Bayona     ACTOR   
85576       85576        In??s Prieto                    Fanny     ACTOR   
85577       85577        Isabel Gaona                   Cacica     ACTOR   
85578       85578      Julian Gaviria                  unknown  DIRECTOR   

             title       type  release_year              genres  imdb_score  \
0      Taxi Driver      MOVIE          1976  ['drama', 'crime']         8.2   
1      Taxi Driver      MOVIE          1976  ['drama', 'crime']         8.2   
2      Taxi Driver      MOVIE          1976  ['drama', 'crime']         8.2   
3      Taxi Driver      MOVIE          1976  ['drama', 'crime']         8.2   
4      Taxi Driver      MOVIE          1976  ['drama', 'crime']         8.2   
...            ...        ...           ...                 ...         ...   
85574      Lokillo  the movie          2021          ['comedy']         3.8   
85575      Lokillo  the movie          2021          ['comedy']         3.8   
85576      Lokillo  the movie          2021          ['comedy']         3.8   
85577      Lokillo      MOVIE          2021          ['comedy']         3.8   
85578      Lokillo  the movie          2021          ['comedy']         3.8   

       imdb_votes  
0        808582.0  
1        808582.0  
2        808582.0  
3        808582.0  
4        808582.0  
...           ...  
85574        68.0  
85575        68.0  
85576        68.0  
85577        68.0  
85578        68.0  

[80853 rows x 10 columns]

Make sure the table doesn't contain any more missing values. Count the missing values again.

In [11]:
# counting missing values
new_table = df.dropna()
new_table.isna().sum()

unnamed:_o      0
name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [12]:
# counting duplicate rows
dup = df.dropna()
print(f'There are {dup.duplicated().sum()} duplicates.')

There are 0 duplicates.


Review the duplicate rows to determine if removing them would distort our dataset.

In [13]:
# Produce table with duplicates (with original rows included) and review last 5 rows
copy = dup[dup.duplicated()]

copy.tail(5)

Empty DataFrame
Columns: [unnamed:_o, name, character, role, title, type, release_year, genres, imdb_score, imdb_votes]
Index: []

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [14]:
# removing duplicate rows
df = df.drop_duplicates()
df = df.reset_index()

df.tail(5)

index  unnamed:_o                name     character      role    title  \
85574  85574       85574    Adelaida Buscato    Mar??a Paz     ACTOR  Lokillo   
85575  85575       85575  Luz Stella Luengas  Karen Bayona     ACTOR  Lokillo   
85576  85576       85576        In??s Prieto         Fanny     ACTOR  Lokillo   
85577  85577       85577        Isabel Gaona        Cacica     ACTOR  Lokillo   
85578  85578       85578      Julian Gaviria       unknown  DIRECTOR  Lokillo   

            type  release_year      genres  imdb_score  imdb_votes  
85574  the movie          2021  ['comedy']         3.8        68.0  
85575  the movie          2021  ['comedy']         3.8        68.0  
85576  the movie          2021  ['comedy']         3.8        68.0  
85577      MOVIE          2021  ['comedy']         3.8        68.0  
85578  the movie          2021  ['comedy']         3.8        68.0

Check for duplicate rows once more to make sure you have removed all of them:

In [15]:
# checking for duplicates
df = df.dropna()
print(f'There are {df.duplicated().sum()} duplicates.')

There are 0 duplicates.


Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [16]:
# viewing unique type names
print(sorted(df['type'].unique()))

['MOVIE', 'SHOW', 'movies', 'shows', 'the movie', 'tv', 'tv series', 'tv show', 'tv shows']


Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters:
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [17]:
# function for replacing implicit duplicates
def replace_wrong_show(wrong_shows_list, correct_show):
    for wrong_show in wrong_shows_list:
        df['type'].replace(wrong_show, correct_show, inplace=True)

wrong_shows_list = ['SHOW', 'shows', 'tv show', 'tv shows', 'tv series', 'tv']
correct_show = 'show'

Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [18]:
# removing implicit duplicates
replace_wrong_show(wrong_shows_list, correct_show)

Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [19]:
# viewing unique genre names
df['type'].unique()

array(['MOVIE', 'the movie', 'show', 'movies'], dtype=object)

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed.

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [20]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
# give the slice of dataframe new name
golden_age = df[df['release_year'] >= 1999]

In [21]:
# repeat conditional indexing so df has only shows (movies are removed as result)
golden_age = golden_age[golden_age['type'] == 'show']

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [22]:
# rounding column with scores
#checking the outcome with tail()
golden_age['imdb_score'] = golden_age['imdb_score'].round()
# how to round numbers: https://www.w3schools.com/python/pandas/ref_df_round.asp

golden_age.tail(5)

index  unnamed:_o                 name character      role      title  \
85433  85433       85433     Maneerat Kam-Uan        Ae     ACTOR  Let's Eat   
85434  85434       85434    Rudklao Amratisha   unknown     ACTOR  Let's Eat   
85435  85435       85435      Jaturong Mokjok   unknown     ACTOR  Let's Eat   
85436  85436       85436     Pisamai Wilaisak   unknown     ACTOR  Let's Eat   
85437  85437       85437  Sarawut Wichiensarn   unknown  DIRECTOR  Let's Eat   

       type  release_year               genres  imdb_score  imdb_votes  
85433  show          2021  ['drama', 'comedy']         8.0         5.0  
85434  show          2021  ['drama', 'comedy']         8.0         5.0  
85435  show          2021  ['drama', 'comedy']         8.0         5.0  
85436  show          2021  ['drama', 'comedy']         8.0         5.0  
85437  show          2021  ['drama', 'comedy']         8.0         5.0

It is now time to identify outliers based on the number of votes.

In [23]:
# Use groupby() for scores and count all unique values in each group, print the result
print(golden_age.groupby('imdb_score')['imdb_votes'].count())

# Coding in jupyter notebook is interesting, I'm used to coding in VS Code and visually that's what caused me
# to code this the way I initially did, I finally understand, wasnt until this error cause the next error

imdb_score
2.0       40
3.0       27
4.0      225
5.0      691
6.0     2986
7.0     5322
8.0     5200
9.0      583
10.0       8
Name: imdb_votes, dtype: int64


Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [24]:
# filter dataframe using two conditions (scores to be in the range 4-9)
# group scores and corresponding average number of votes, reset index and print the result

golden_age = golden_age.groupby('imdb_score')['imdb_votes'].mean()
# how to do the mean:
# https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.mean.html

golden_age = golden_age.iloc[2:8].reset_index()
# set the range for index:
# https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html

Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [25]:
# round column with averages
# rename columns
# print dataframe in descending order
golden_age.rename(columns={'imdb_score': 'avg_score', 'imdb_votes': 'avg_votes'}).sort_values(by='avg_score', ascending=False).reset_index()

index  avg_score      avg_votes
0      5        9.0  121401.017153
1      4        8.0   28361.942115
2      3        7.0    8012.183953
3      2        6.0    3168.731413
4      1        5.0    2846.192475
5      0        4.0    6544.480000

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.